File: client/Hysj.Client/Crypto/Ratchet/DoubleRatchet.cs
using System.Security.Cryptography;
using System.Text;

namespace Hysj.Client.Crypto.Ratchet;

/// <summary>
/// Signal Double Ratchet implementasjon.
/// Gir forward secrecy per melding.
/// </summary>
public sealed class DoubleRatchet
{
    private RatchetState _state;

    private DoubleRatchet(RatchetState state) => _state = state;

    /// <summary>Initialiser som avsender (Alice) etter X3DH handshake.</summary>
    public static DoubleRatchet InitSender(byte[] sharedSecret, byte[] bobDhPublic)
    {
        var (rootKey, chainKey) = KdfRk(sharedSecret, new byte[32]);
        var sendDh = EccKeyPair.Generate();
        var (newRoot, sendChain) = KdfRk(rootKey, sendDh.DeriveSharedSecret(bobDhPublic));

        return new DoubleRatchet(new RatchetState
        {
            RootKey              = newRoot,
            SendingChainKey      = sendChain,
            ReceivingChainKey    = chainKey,
            DHSendPrivate        = sendDh.ExportPrivateKey(),
            DHSendPublic         = sendDh.ExportPublicKey(),
            DHReceivePublic      = bobDhPublic,
        });
    }

    /// <summary>Initialiser som mottaker (Bob) etter X3DH handshake.</summary>
    public static DoubleRatchet InitReceiver(byte[] sharedSecret, byte[] myDhPrivate)
    {
        var (rootKey, chainKey) = KdfRk(sharedSecret, new byte[32]);
        var myDh = EccKeyPair.FromPrivateKey(myDhPrivate);

        return new DoubleRatchet(new RatchetState
        {
            RootKey           = rootKey,
            ReceivingChainKey = chainKey,
            DHSendPrivate     = myDh.ExportPrivateKey(),
            DHSendPublic      = myDh.ExportPublicKey(),
        });
    }

    public static DoubleRatchet FromState(RatchetState state) => new(state);

    public RatchetState ExportState() => _state;

    /// <summary>Krypter en melding. Returnerer kryptert blob + header.</summary>
    public (byte[] ciphertext, MessageHeader header) Encrypt(byte[] plaintext)
    {
        var (newChain, msgKey) = KdfCk(_state.SendingChainKey);
        _state.SendingChainKey = newChain;

        using var mk = new MessageKey(msgKey);
        var cipher = AesGcmCipher.Encrypt(plaintext, mk.Key);

        var header = new MessageHeader(
            _state.DHSendPublic,
            _state.SendingIndex,
            _state.PreviousSendingLength);

        _state.SendingIndex++;
        return (cipher, header);
    }

    /// <summary>Dekrypter en melding.</summary>
    public byte[] Decrypt(byte[] ciphertext, MessageHeader header)
    {
        // Sjekk hoppet-over nøkler
        var skipKey = $"{Convert.ToBase64String(header.DHPublic)}:{header.MessageIndex}";
        if (_state.SkippedKeys.TryGetValue(skipKey, out var skipped))
        {
            _state.SkippedKeys.Remove(skipKey);
            using var sk = new MessageKey(skipped);
            return AesGcmCipher.Decrypt(ciphertext, sk.Key);
        }

        // DH ratchet hvis ny nøkkel fra motpart
        if (!header.DHPublic.SequenceEqual(_state.DHReceivePublic))
        {
            SkipMessageKeys(_state.ReceivingChainKey, header.PreviousChainLength);
            PerformDhRatchet(header.DHPublic);
        }

        SkipMessageKeys(_state.ReceivingChainKey, header.MessageIndex);

        var (newChain, msgKey) = KdfCk(_state.ReceivingChainKey);
        _state.ReceivingChainKey = newChain;
        _state.ReceivingIndex++;

        using var mk = new MessageKey(msgKey);
        return AesGcmCipher.Decrypt(ciphertext, mk.Key);
    }

    private void PerformDhRatchet(byte[] theirDhPublic)
    {
        _state.PreviousSendingLength = _state.SendingIndex;
        _state.SendingIndex          = 0;
        _state.ReceivingIndex        = 0;
        _state.DHReceivePublic       = theirDhPublic;

        using var oldDh = EccKeyPair.FromPrivateKey(_state.DHSendPrivate);
        var dh = oldDh.DeriveSharedSecret(theirDhPublic);
        (_state.RootKey, _state.ReceivingChainKey) = KdfRk(_state.RootKey, dh);

        using var newDh = EccKeyPair.Generate();
        _state.DHSendPrivate = newDh.ExportPrivateKey();
        _state.DHSendPublic  = newDh.ExportPublicKey();

        var dh2 = newDh.DeriveSharedSecret(theirDhPublic);
        (_state.RootKey, _state.SendingChainKey) = KdfRk(_state.RootKey, dh2);

        CryptographicOperations.ZeroMemory(dh);
        CryptographicOperations.ZeroMemory(dh2);
    }

    private void SkipMessageKeys(byte[] chainKey, int until)
    {
        const int MaxSkip = 1000;
        var ck = chainKey;
        var idx = _state.ReceivingIndex;
        while (idx < until && _state.SkippedKeys.Count < MaxSkip)
        {
            var (newCk, mk) = KdfCk(ck);
            ck = newCk;
            var key = $"{Convert.ToBase64String(_state.DHReceivePublic)}:{idx}";
            _state.SkippedKeys[key] = mk;
            idx++;
        }
    }

    // KDF for root chain: returns (new root key, chain key)
    private static (byte[] rootKey, byte[] chainKey) KdfRk(byte[] rk, byte[] dh)
    {
        var derived = HkdfDeriver.DeriveKey(dh, rk, "hysj-ratchet-root", 64);
        return (derived[..32], derived[32..]);
    }

    // KDF for message chain: returns (new chain key, message key)
    private static (byte[] chainKey, byte[] messageKey) KdfCk(byte[] ck)
    {
        var mk = HMACSHA256(ck, [0x01]);
        var nk = HMACSHA256(ck, [0x02]);
        return (nk, mk);
    }

    private static byte[] HMACSHA256(byte[] key, byte[] data)
    {
        using var hmac = new System.Security.Cryptography.HMACSHA256(key);
        return hmac.ComputeHash(data);
    }
}

File: hysj-app/src/crypto/ratchet/doubleRatchet.ts
/**
 * Signal Double Ratchet implementation.
 * Provides forward secrecy per message: each message uses a unique key
 * that is deleted immediately after use.
 *
 * Fixes over the MAUI version:
 * - SkipMessageKeys correctly writes back the updated chain key
 * - Uses cryptographic random for all key generation
 */
import { generateKeyPair, deriveSharedSecret, zeroMemory } from '../keys';
import { hkdfDeriveKey, hmacSha256 } from '../kdf';
import { encrypt, decrypt } from '../cipher';
import type { RatchetState, MessageHeader, EncryptedMessage } from './types';
import { toBase64 } from '../encoding';

const MAX_SKIP = 1000;

/** Initialize as sender (Alice) after X3DH handshake. */
export function initSender(sharedSecret: Uint8Array, bobDhPublic: Uint8Array): RatchetState {
  const [rootKey, chainKey] = kdfRk(sharedSecret, new Uint8Array(32));
  const sendDh = generateKeyPair();
  const dhOut = deriveSharedSecret(sendDh.secretKey, bobDhPublic);
  const [newRoot, sendChain] = kdfRk(rootKey, dhOut);
  zeroMemory(dhOut);

  return {
    rootKey: newRoot,
    sendingChainKey: sendChain,
    receivingChainKey: chainKey,
    dhSendSecret: sendDh.secretKey,
    dhSendPublic: sendDh.publicKey,
    dhReceivePublic: bobDhPublic,
    sendingIndex: 0,
    receivingIndex: 0,
    previousSendingLength: 0,
    skippedKeys: {},
  };
}

/** Initialize as receiver (Bob) after X3DH handshake. */
export function initReceiver(sharedSecret: Uint8Array, myDhSecret: Uint8Array, myDhPublic: Uint8Array): RatchetState {
  const [rootKey, chainKey] = kdfRk(sharedSecret, new Uint8Array(32));

  return {
    rootKey,
    sendingChainKey: new Uint8Array(32),
    receivingChainKey: chainKey,
    dhSendSecret: myDhSecret,
    dhSendPublic: myDhPublic,
    dhReceivePublic: new Uint8Array(32),
    sendingIndex: 0,
    receivingIndex: 0,
    previousSendingLength: 0,
    skippedKeys: {},
  };
}

/** Encrypt a plaintext message. Mutates state. */
export function ratchetEncrypt(state: RatchetState, plaintext: Uint8Array): EncryptedMessage {
  const [newChain, msgKey] = kdfCk(state.sendingChainKey);
  state.sendingChainKey = newChain;

  const ciphertext = encrypt(plaintext, msgKey);
  zeroMemory(msgKey);

  const header: MessageHeader = {
    dhPublic: state.dhSendPublic,
    messageIndex: state.sendingIndex,
    previousChainLength: state.previousSendingLength,
  };
  state.sendingIndex++;

  return { ciphertext, header };
}

/** Decrypt a message. Mutates state. */
export function ratchetDecrypt(state: RatchetState, msg: EncryptedMessage): Uint8Array {
  const { ciphertext, header } = msg;

  // Check skipped keys first
  const skipKey = `${toBase64(header.dhPublic)}:${header.messageIndex}`;
  const skipped = state.skippedKeys[skipKey];
  if (skipped) {
    delete state.skippedKeys[skipKey];
    const plaintext = decrypt(ciphertext, skipped);
    zeroMemory(skipped);
    return plaintext;
  }

  // DH ratchet if new key from peer
  if (!uint8Eq(header.dhPublic, state.dhReceivePublic)) {
    skipMessageKeys(state, header.previousChainLength);
    performDhRatchet(state, header.dhPublic);
  }

  skipMessageKeys(state, header.messageIndex);

  const [newChain, msgKey] = kdfCk(state.receivingChainKey);
  state.receivingChainKey = newChain;
  state.receivingIndex++;

  const plaintext = decrypt(ciphertext, msgKey);
  zeroMemory(msgKey);
  return plaintext;
}

// --- Internal helpers ---

function performDhRatchet(state: RatchetState, theirDhPublic: Uint8Array): void {
  state.previousSendingLength = state.sendingIndex;
  state.sendingIndex = 0;
  state.receivingIndex = 0;
  state.dhReceivePublic = theirDhPublic;

  const dh1 = deriveSharedSecret(state.dhSendSecret, theirDhPublic);
  [state.rootKey, state.receivingChainKey] = kdfRk(state.rootKey, dh1);
  zeroMemory(dh1);

  const newDh = generateKeyPair();
  zeroMemory(state.dhSendSecret);
  state.dhSendSecret = newDh.secretKey;
  state.dhSendPublic = newDh.publicKey;

  const dh2 = deriveSharedSecret(newDh.secretKey, theirDhPublic);
  [state.rootKey, state.sendingChainKey] = kdfRk(state.rootKey, dh2);
  zeroMemory(dh2);
}

/**
 * Skip message keys up to `until` index and store them for later.
 * FIX: correctly writes back the updated receiving chain key.
 */
function skipMessageKeys(state: RatchetState, until: number): void {
  let ck = state.receivingChainKey;
  while (state.receivingIndex < until && Object.keys(state.skippedKeys).length < MAX_SKIP) {
    const [newCk, mk] = kdfCk(ck);
    const key = `${toBase64(state.dhReceivePublic)}:${state.receivingIndex}`;
    state.skippedKeys[key] = mk;
    ck = newCk;
    state.receivingIndex++;
  }
  // Write back the updated chain key
  state.receivingChainKey = ck;
}

/** KDF for root chain: returns [new root key, chain key]. */
function kdfRk(rk: Uint8Array, dh: Uint8Array): [Uint8Array, Uint8Array] {
  const derived = hkdfDeriveKey(dh, rk, 'hysj-ratchet-root', 64);
  return [derived.slice(0, 32), derived.slice(32)];
}

/** KDF for message chain: returns [new chain key, message key]. */
function kdfCk(ck: Uint8Array): [Uint8Array, Uint8Array] {
  const mk = hmacSha256(ck, new Uint8Array([0x01]));
  const nk = hmacSha256(ck, new Uint8Array([0x02]));
  return [nk, mk];
}

function uint8Eq(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) diff |= a[i] ^ b[i];
  return diff === 0;
}
